transaction/core/src/tx.rs

```diff
@@ -262,26 +268,92 @@ impl TxOut {
         tx_private_key: &RistrettoPrivate,
         hint: EncryptedFogHint,
     ) -> Result<Self, AmountError> {
+        TxOut::new_with_memo(value, recipient, tx_private_key, hint, |_| {
+            Ok(MemoPayload::default())
+        })
+        .map_err(|err| match err {
+            NewTxError::Amount(err) => err,
+            NewTxError::Memo(_) => unreachable!(),
+        })
+    }
+
+    /// Creates a TxOut that sends `value` to `recipient`, with a custom memo
+    /// attached. The memo is produced by a callback function which is
+    /// passed the value and tx_public_key.
+    ///
+    /// # Arguments
+    /// * `value` - Value of the output.
+    /// * `recipient` - Recipient's address.
+    /// * `tx_private_key` - The transaction's private key
+    /// * `hint` - Encrypted Fog hint.
+    /// * `memo_fn` - A callback taking tx_public_key, which produces a
+    ///   MemoPayload, or a String error
+    pub fn new_with_memo(
+        value: u64,
+        recipient: &PublicAddress,
+        tx_private_key: &RistrettoPrivate,
+        hint: EncryptedFogHint,
+        memo_fn: impl FnOnce(&RistrettoPublic) -> Result<MemoPayload, String>,
```

Comment:
I wonder if passing only the `tx_public_key` as context for generating the memo is not too specific for our particular use case. The comment at the beginning of the method says the value is also passed.
My suggestion would be to pass as much context as we can, so future users have it in case they want it.
I think passing the value and recipient would be nice.

Reply:
So the way I got here was:
- I knew that I needed at least the tx_public_key because that goes into the MAC in the authenticated memos
- The value is probably important for the change tx out
- Other stuff?

So I started with just two things as memo function arguments, and plumbed that through the various TransactionBuilder calls. At some later point, I figured out that the MemoBuilder API should be invoked once for each AddOutput and AddChangeOutput call, and I had some idea what state that should have and what that should look like.

Then I got to this point in the code, where we're going to make the `memo_fn` be something that invokes the `MemoBuilder`:

Reply:
The only reason we really need the `memo_fn` to be a callback and not just a `MemoPayload` like before, is that some things like `tx_public_key` are created in `TxOut::new` immediately before the memo is constructed, but the memo contents want to make that part of their HMAC and such.

So, I think we could pass this callback all of those things on the stack (in this function below), so like, the target key, the amount, maybe even the shared secret?

I thought it was kind of nice that `TxOut::new` is managing the encryption of the memo, and I think right now, the `shared_secret` is usually discarded after `TxOut::new` so doing that encryption call anywhere else would require recomputation or some messy stuff to preserve the `shared_secret`.

Reply:
What I'm thinking now is, maybe there should be a struct like

```
pub struct MemoContext<'a> {
    tx_public_key: &'a RistrettoPublic,
}
```

and that thing is the argument to `memo_fn`.
And if we need to add more arguments to this callback later, it's not a very disruptive change to add them to this struct, and it's fairly readable.

For stuff like value and public address, that could be captured by the caller, I think we don't need to make those part of `MemoContext`, I'd rather have that be only the values that aren't even computed until `TxOut::new` has been called

Reply:
The struct is a good idea.